### ISSUE TITLE
Exception when sorting warning messages if a warning has no line number

### ISSUE THREAD

**Issue by kovidgoyal** [sphinx-doc/sphinx#10498]
### Describe the bug

In sphinx 5.0 there is a new warning about extlinks. This warning has no associated linenumber. This breaks Catalog.__iter__ with the following traceback:

```
# Sphinx version: 5.0.0
# Python version: 3.10.4 (CPython)
# Docutils version: 0.18.1 release
# Jinja2 version: 3.1.2
# Last messages:
#   writing output... [ 88%] template_lang
#   writing output... [ 90%] tutorials
#   writing output... [ 92%] typesetting_math
#   writing output... [ 94%] url_scheme
#   writing output... [ 96%] viewer
#   writing output... [ 98%] virtual_libraries
#   writing output... [100%] xpath
#   
#   writing message catalogs... [  2%] sphinx
#   writing message catalogs... [  5%] catalogs
# Loaded extensions:
#   sphinx.ext.mathjax (5.0.0) from /usr/lib/python3.10/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from /usr/lib/python3.10/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from /usr/lib/python3.10/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (2.0.0) from /usr/lib/python3.10/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.5) from /usr/lib/python3.10/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from /usr/lib/python3.10/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /usr/lib/python3.10/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc.preserve_defaults (1.0) from /usr/lib/python3.10/site-packages/sphinx/ext/autodoc/preserve_defaults.py
#   sphinx.ext.autodoc.type_comment (5.0.0) from /usr/lib/python3.10/site-packages/sphinx/ext/autodoc/type_comment.py
#   sphinx.ext.autodoc (5.0.0) from /usr/lib/python3.10/site-packages/sphinx/ext/autodoc/__init__.py
#   custom (unknown version) from /home/kovid/work/calibre/manual/custom.py
#   sidebar_toc (unknown version) from /home/kovid/work/calibre/manual/sidebar_toc.py
#   sphinx.ext.viewcode (5.0.0) from /usr/lib/python3.10/site-packages/sphinx/ext/viewcode.py
#   sphinx.ext.extlinks (5.0.0) from /usr/lib/python3.10/site-packages/sphinx/ext/extlinks.py
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/sphinx/cmd/build.py", line 276, in build_main
    app.build(args.force_all, filenames)
  File "/usr/lib/python3.10/site-packages/sphinx/application.py", line 329, in build
    self.builder.build_update()
  File "/usr/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 288, in build_update
    self.build(to_build,
  File "/usr/lib/python3.10/site-packages/sphinx/builders/gettext.py", line 251, in build
    super().build(docnames, summary, method)
  File "/usr/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 355, in build
    self.finish()
  File "/usr/lib/python3.10/site-packages/sphinx/builders/gettext.py", line 273, in finish
    context['messages'] = list(catalog)
  File "/usr/lib/python3.10/site-packages/sphinx/builders/gettext.py", line 60, in __iter__
    positions = sorted(set((source, line) for source, line, uuid
TypeError: '<' not supported between instances of 'int' and 'NoneType'
```

The fix is simply to change the  sorting of positions to:
```py
            positions = sorted(set((source, line or 0) for source, line, uuid
                                   in self.metadata[message]))
```

### How to Reproduce

Use extlinks with sphinx 5

### Your project

calibre

### OS

Linux

### Python version

3.10.4

### Sphinx version

5.0.0

**Comment by kovidgoyal** [comment]
Actually it wasnt the extlinks warning. Instead if you have an image with an alt field that has exactly the same text as a preceding heading, you get the message with None as a linenumber.

**Comment by kovidgoyal** [comment]
Here is the commit I made to workaround it in my project: https://github.com/kovidgoyal/calibre/commit/bece99941d6a8c4a708c3950791ac4d22bcffb6d

**Comment by AA-Turner** [comment]
@kovidgoyal do you have a single file reproducer please?

A

**Comment by AA-Turner** [comment]
Also, your commit message says "Workaround yet another regression in Sphinx" -- what are the other regressions? Can we fix them? We'd like for upgrading Sphinx versions to be as painless as possible, so any help would be really useful here.

Thanks,
Adam

**Comment by AA-Turner** [comment]
If I run the reproducer file here:

```python
import shutil
from pathlib import Path

from sphinx.cmd.make_mode import run_make_mode

def write(filename, text): Path(filename).write_text(text, encoding="utf-8")


write("conf.py", '''\
extensions = ['sphinx.ext.extlinks']
''')

write("index.rst", '''\
lorem ipsum dolar sit amet
--------------------------

.. image:: placeholder_image.jpg
   :alt: lorem ipsum dolar sit amet
''')

shutil.rmtree("_build", ignore_errors=True)
run_make_mode(["html", ".", "_build", "-T"])
```

as `python reproducer_10498.py`, I don't get the error. Are you building for a different language?

A

**Comment by kovidgoyal** [comment]
touch conf.py
sphinx-build -b gettext -t online -t gettext . /tmp/x

**Comment by AA-Turner** [comment]
Docutils doesn't set a line number for image nodes with alt text for some reason, reported as [bug#449](https://sourceforge.net/p/docutils/bugs/449/)

We could patch Docutils, or add the workaround you suggested -- I'm inclined to go with the latter as it is more general (although might hide similar problems in the future)

A

**Comment by kovidgoyal** [comment]
On Tue, May 31, 2022 at 10:22:14AM -0700, Adam Turner wrote:
> Docutils doesn't set a line number for image nodes with alt text for some reason, reported as [bug#449](https://sourceforge.net/p/docutils/bugs/449/)

Its not image nodes with alt text but image nodes with alt text that is
the same as the text of a previous title.

> 
> We could patch Docutils, or add the workaround you suggested -- I'm inclined to go with the latter as it is more general (although might hide similar problems in the future)

A proper workaround is to check for no line number and issue a warning
and then use the sort fix I suggested.

**Comment by AA-Turner** [comment]
```pycon
>>> print(core.publish_doctree(".. image:: placeholder_image.jpg")[0].line)
1
>>> print(core.publish_doctree(".. image:: placeholder_image.jpg\n   :alt: blah")[0].line)
None
```

This example has no title, yet still fails.

A

**Comment by AA-Turner** [comment]
> A proper workaround is to check for no line number and issue a warning and then use the sort fix I suggested.

The PR has been updated to issue a warning, thanks for the suggestion.

A

**Comment by tk0miya** [comment]
Fixed by #10503. It will be released as v5.0.1 soon (within a few days). Please wait for a while.
Thank you for your report.
